dids/did_dht.go

```diff
@@ -0,0 +1,310 @@
+	"del":  {},
+}
+
+var relationshipDNStoDID = map[string]string{
+	"auth": "authentication",
+	"asm":  "assertionMethod",
+	"agm":  "keyAgreement",
+	"inv":  "capabilityInvocation",
+	"del":  "capabilityDelegation",
+}
+
+var keyTypes = map[string]string{
+	"0": dsa.AlgorithmIDED25519,
+	"1": dsa.AlgorithmIDSECP256K1,
+	"2": dsa.AlgorithmIDSECP256K1,
+}
+
+// dhtDIDRecord is used to structure the DNS representation of a DID
+type dhtDIDRecord struct {
+	rootRecord string
+	records    map[string]string
+}
+
+func (rec *dhtDIDRecord) DIDDocument(didURI string) *Document {
+	relationshipMap := parseVerificationRelationships(rec.rootRecord)
+
+	// Now we have a did in a dns record. yay
+	document := &Document{
+		ID: didURI,
+	}
+
+	// Now create the did document
+	for name, data := range rec.records {
+
+		switch {
+		case strings.HasPrefix(name, "_k"):
+			{
+				vMethod, err := UnmarshalVerificationMethod(data)
+				if err != nil {
+					// TODO handle error
+				}
+
+				// extracting kN from _kN._did
+				entryId := strings.Split(name, ".")[0][1:]
+				relationships, ok := relationshipMap[entryId]
+
+				if !ok {
+					// no relationships
+					continue
+				}
+
+				opts := []string{}
+				for _, r := range relationships {
+					if o, ok := relationshipDNStoDID[r]; ok {
+						opts = append(opts, o)
+					}
+				}
+
+				document.AddVerificationMethod(
+					*vMethod,
+					Purposes(opts...),
+				)
+			}
+		case strings.HasPrefix(name, "_s"):
+			s := UnmarshalService(data)
+			document.AddService(s)
+		case strings.HasPrefix(name, "_cnt"):
+			// TODO add controller https://did-dht.com/#controller
+			// optional field
+			// comma-separated list of controller DID identifiers
+			document.Controller = strings.Split(data, ",")
+		case strings.HasPrefix(name, "_aka"):
+			// TODO add aka https://did-dht.com/#also-known-as
+			document.AlsoKnownAs = strings.Split(data, ",")
+		default:
+		}
+	}
+
+	return document
+}
+
+func ResolveDIDDHT(uri, relay string, client *http.Client) (*Document, error) {
+	// 1. Parse URI and make sure it's a DHT method
+	// 2. Get the public key bytes / decode did.id from z-base-32
+	// 3. create the bep44 message - needs the zbase32 encoded publicKey / did.id
+	// 4. create the Pkarr url with the relay + did.id
+	// 5. fetch
+	// 6. bep44 message using (publicKey bytes - decoded zbase32 - ) etc.
+	// 7. fetch the dns packet using the bep44 message
+	// 8. decode the dns packet
+	// 9. create the did document
+
+	// 1. Parse URI and make sure it's a DHT method
+	did, err := Parse(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	// 2. ensure did ID is zbase32
+	identifier, err := zbase32.DecodeString(did.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(identifier) <= 0 {
+		return nil, fmt.Errorf("no bytes decoded from zbase32 identifier %s", did.ID)
+	}
+
+	// 3. fetch bep44 encoded did document
+	res, err := client.Get(fmt.Sprintf("%s/%s", relay, did.ID))
+	if err != nil {
+		return nil, err
+	}
```

Comment:
should we wrap this error with additional context? also sry, dumb q, does `Get` return an error if status code is >= 4xx?

Reply:
yes, we should make the call with a context. I don't know what additional errors might happen other than request failures. I don't know if we can get back a <400 but with some body that signifies an error. Need to look at the DHT spec itself I'm guessing